Create a `aarch64-apple-ios-sim` target for `rustc`

Currently in our application we cannot run the iOS simulator on M1 machines because it is expecting `x86_64` architecture. Forcing `arm64` using the `rustc` target `aarch64-apple-ios` can build the project and ship the application but this targets the iOS Device SDK, and will not build in the simulator for testing with this error: 

```
ld: <path/to/lib> building for iOS Simulator, but linking in object file built for iOS, for architecture arm64
```

`rustc` should provide an `aarch-apple-ios-sim` target that uses the iOS Simulator SDK.

I built the toolchain successfully, but when I try to compile my project (with either `aarch64-apple-darwin` or `aarch64-apple-ios-sim`) I'm getting: 

```
error[E0277]: cannot subtract `usize` from `u32`
    --> $CARGO_REGISTRY_PATH/lexical-core-0.7.4/src/atof/algorithm/math.rs:2065:25
     |
2065 |     let rs = Limb::BITS - s;
     |                         ^ no implementation for `u32 - usize`
     |
     = help: the trait `Sub<usize>` is not implemented for `u32`
 ```
 
 I think I've missed a setting in the build config I'm using, because the normal nightly compiles fine :( 
 
 `config.toml`: https://gist.github.com/deg4uss3r/183ef399590f74353691b21badc382cc

@badboy Yeah I added the patch to my workspace for now and I'm getting somewhere but still failing with `clang`: 

```
 Ld /Users/rth/Library/Developer/Xcode/DerivedData/1Password-dabitbybocruqiamiqeqnvlwnpmd/Build/Intermediates.noindex/1PasswordFrameworks.build/Debug-iphonesimulator/Core-iOS.build/Objects-normal/x86_64/Binary/OnePasswordCore normal x86_64 (in target 'Core-iOS' from project '1PasswordFrameworks')
    cd /Users/rth/work/core/apple/1PasswordFrameworks
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -target x86_64-apple-ios14.0-simulator -dynamiclib -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator14.4.sdk -L/Users/rth/Library/Developer/Xcode/DerivedData/1Password-dabitbybocruqiamiqeqnvlwnpmd/Build/Products/Debug-iphonesimulator -L/Users/rth/work/core/apple/1PasswordFrameworks/../../target/aarch64-apple-ios-sim/debug -F/Users/rth/Library/Developer/Xcode/DerivedData/1Password-dabitbybocruqiamiqeqnvlwnpmd/Build/Products/Debug-iphonesimulator -filelist /Users/rth/Library/Developer/Xcode/DerivedData/1Password-dabitbybocruqiamiqeqnvlwnpmd/Build/Intermediates.noindex/1PasswordFrameworks.build/Debug-iphonesimulator/Core-iOS.build/Objects-normal/x86_64/OnePasswordCore.LinkFileList -install_name @rpath/OnePasswordCore.framework/OnePasswordCore -Xlinker -rpath -Xlinker @executable_path/Frameworks -Xlinker -rpath -Xlinker @loader_path/Frameworks -Xlinker -rpath -Xlinker @executable_path/Frameworks -Xlinker -rpath -Xlinker @loader_path/Frameworks -dead_strip -Xlinker -object_path_lto -Xlinker /Users/rth/Library/Developer/Xcode/DerivedData/1Password-dabitbybocruqiamiqeqnvlwnpmd/Build/Intermediates.noindex/1PasswordFrameworks.build/Debug-iphonesimulator/Core-iOS.build/Objects-normal/x86_64/OnePasswordCore_lto.o -Xlinker -export_dynamic -Xlinker -no_deduplicate -Xlinker -objc_abi_version -Xlinker 2 -fapplication-extension -fobjc-link-runtime -L/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/iphonesimulator -L/usr/lib/swift -Xlinker -add_ast_path -Xlinker /Users/rth/Library/Developer/Xcode/DerivedData/1Password-dabitbybocruqiamiqeqnvlwnpmd/Build/Intermediates.noindex/1PasswordFrameworks.build/Debug-iphonesimulator/Core-iOS.build/Objects-normal/x86_64/OnePasswordCore.swiftmodule -lonepasswordcore -Xlinker -no_adhoc_codesign -compatibility_version 1 -current_version 1 -Xlinker -dependency_info -Xlinker /Users/rth/Library/Developer/Xcode/DerivedData/1Password-dabitbybocruqiamiqeqnvlwnpmd/Build/Intermediates.noindex/1PasswordFrameworks.build/Debug-iphonesimulator/Core-iOS.build/Objects-normal/x86_64/OnePasswordCore_dependency_info.dat -o /Users/rth/Library/Developer/Xcode/DerivedData/1Password-dabitbybocruqiamiqeqnvlwnpmd/Build/Intermediates.noindex/1PasswordFrameworks.build/Debug-iphonesimulator/Core-iOS.build/Objects-normal/x86_64/Binary/OnePasswordCore

ld: warning: ignoring file /Users/rth/work/core/target/aarch64-apple-ios-sim/debug/libonepasswordcore.a, building for iOS Simulator-x86_64 but attempting to link with file built for unknown-arm64
Undefined symbols for architecture x86_64:
```

I tried doing this on my own (I have no idea how Xcode is getting this clang call from).. 

with: 

```
❯ /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -target arm64-apple-ios14.0-simulator -dynamiclib -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator14.4.sdk -L/Users/rth/Library/Developer/Xcode/DerivedData/1Password-dabitbybocruqiamiqeqnvlwnpmd/Build/Products/Debug-iphonesimulator -L/Users/rth/work/core/apple/1PasswordFrameworks/../../target/aarch64-apple-ios-sim/debug -F/Users/rth/Library/Developer/Xcode/DerivedData/1Password-dabitbybocruqiamiqeqnvlwnpmd/Build/Products/Debug-iphonesimulator -filelist /Users/rth/Library/Developer/Xcode/DerivedData/1Password-dabitbybocruqiamiqeqnvlwnpmd/Build/Intermediates.noindex/1PasswordFrameworks.build/Debug-iphonesimulator/Core-iOS.build/Objects-normal/arm64/OnePasswordCore.LinkFileList -install_name @rpath/OnePasswordCore.framework/OnePasswordCore -Xlinker -rpath -Xlinker @executable_path/Frameworks -Xlinker -rpath -Xlinker @loader_path/Frameworks -Xlinker -rpath -Xlinker @executable_path/Frameworks -Xlinker -rpath -Xlinker @loader_path/Frameworks -dead_strip -Xlinker -object_path_lto -Xlinker /Users/rth/Library/Developer/Xcode/DerivedData/1Password-dabitbybocruqiamiqeqnvlwnpmd/Build/Intermediates.noindex/1PasswordFrameworks.build/Debug-iphonesimulator/Core-iOS.build/Objects-normal/arm64/OnePasswordCore_lto.o -Xlinker -export_dynamic -Xlinker -no_deduplicate -Xlinker -objc_abi_version -Xlinker 2 -fapplication-extension -fobjc-link-runtime -L/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/iphonesimulator -L/usr/lib/swift -Xlinker -add_ast_path -Xlinker /Users/rth/Library/Developer/Xcode/DerivedData/1Password-dabitbybocruqiamiqeqnvlwnpmd/Build/Intermediates.noindex/1PasswordFrameworks.build/Debug-iphonesimulator/Core-iOS.build/Objects-normal/arm64/OnePasswordCore.swiftmodule -lonepasswordcore -Xlinker -no_adhoc_codesign -compatibility_version 1 -current_version 1 -Xlinker -dependency_info -Xlinker /Users/rth/Library/Developer/Xcode/DerivedData/1Password-dabitbybocruqiamiqeqnvlwnpmd/Build/Intermediates.noindex/1PasswordFrameworks.build/Debug-iphonesimulator/Core-iOS.build/Objects-normal/arm64/OnePasswordCore_dependency_info.dat -o /Users/rth/Library/Developer/Xcode/DerivedData/1Password-dabitbybocruqiamiqeqnvlwnpmd/Build/Intermediates.noindex/1PasswordFrameworks.build/Debug-iphonesimulator/Core-iOS.build/Objects-normal/arm64/Binary/OnePasswordCore
```

and got: 

```
ld: in /Users/rth/work/core/target/aarch64-apple-ios-sim/debug/libonepasswordcore.a(sqlite3.o), building for iOS Simulator, but linking in object file built for iOS, for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

--- a/compiler/rustc_codegen_ssa/src/back/link.rs
+++ b/compiler/rustc_codegen_ssa/src/back/link.rs
@@ -2187,6 +2187,7 @@ fn add_apple_sdk(cmd: &mut dyn Linker, sess: &Session, flavor: LinkerFlavor) {
         ("x86_64", "tvos") => "appletvsimulator",
         ("arm", "ios") => "iphoneos",
         ("aarch64", "ios") if llvm_target.contains("macabi") => "macosx",
+        ("aarch64", "ios") if llvm_target.contains("sim") => "iphonesimulator",
         ("aarch64", "ios") => "iphoneos",
         ("x86", "ios") => "iphonesimulator",
         ("x86_64", "ios") if llvm_target.contains("macabi") => "macosx",

--- a/compiler/rustc_target/src/spec/aarch64_apple_ios_sim.rs
+++ b/compiler/rustc_target/src/spec/aarch64_apple_ios_sim.rs
@@ -3,8 +3,16 @@ use crate::spec::{Target, TargetOptions};
 
 pub fn target() -> Target {
     let base = opts("ios", Arch::Arm64_sim);
+
+    // Clang automatically chooses a more specific target based on
+    // IPHONEOS_DEPLOYMENT_TARGET.
+    // This is required for the simulator target to pick the right
+    // MACH-O commands, so we do too.
+    let arch = "arm64";
+    let llvm_target = super::apple_base::ios_sim_llvm_target(arch);
+
     Target {
-        llvm_target: "arm64-apple-ios-simulator".to_string(),
+        llvm_target: llvm_target,
         pointer_width: 64,
         data_layout: "e-m:o-i64:64-i128:128-n32:64-S128".to_string(),
         arch: "aarch64".to_string(),

--- a/compiler/rustc_target/src/spec/aarch64_apple_ios_sim.rs
+++ b/compiler/rustc_target/src/spec/aarch64_apple_ios_sim.rs
@@ -1,0 +1,31 @@
+use super::apple_sdk_base::{opts, Arch};
+use crate::spec::{Target, TargetOptions};
+
+pub fn target() -> Target {
+    let base = opts("ios", Arch::Arm64_sim);
+    Target {
+        llvm_target: "arm64-apple-ios-simulator".to_string(),
+        pointer_width: 64,
+        data_layout: "e-m:o-i64:64-i128:128-n32:64-S128".to_string(),
+        arch: "aarch64".to_string(),
+        options: TargetOptions {
+            features: "+neon,+fp-armv8,+apple-a7".to_string(),
+            eliminate_frame_pointer: false,
+            max_atomic_width: Some(128),
+            unsupported_abis: super::arm_base::unsupported_abis(),
+            forces_embed_bitcode: true,
+            // Taken from a clang build on Xcode 11.4.1.
+            // These arguments are not actually invoked - they just have
+            // to look right to pass App Store validation.
+            bitcode_llvm_cmdline: "-triple\0\
+                arm64-apple-ios14.0-simulator\0\
+                -emit-obj\0\
+                -disable-llvm-passes\0\
+                -target-abi\0\
+                darwinpcs\0\
+                -Os\0"
+                .to_string(),
+            ..base
+        },
+    }
+}

--- a/compiler/rustc_target/src/spec/apple_base.rs
+++ b/compiler/rustc_target/src/spec/apple_base.rs
@@ -54,14 +54,16 @@ pub fn opts(os: &str) -> TargetOptions {
     }
 }
 
-fn macos_deployment_target() -> (u32, u32) {
-    let deployment_target = env::var("MACOSX_DEPLOYMENT_TARGET").ok();
-    let version = deployment_target
+fn deployment_target(var_name: &str) -> Option<(u32, u32)> {
+    let deployment_target = env::var(var_name).ok();
+    deployment_target
         .as_ref()
         .and_then(|s| s.split_once('.'))
-        .and_then(|(a, b)| a.parse::<u32>().and_then(|a| b.parse::<u32>().map(|b| (a, b))).ok());
+        .and_then(|(a, b)| a.parse::<u32>().and_then(|a| b.parse::<u32>().map(|b| (a, b))).ok())
+}
 
-    version.unwrap_or((10, 7))
+fn macos_deployment_target() -> (u32, u32) {
+    deployment_target("MACOSX_DEPLOYMENT_TARGET").unwrap_or((10, 7))
 }
 
 pub fn macos_llvm_target(arch: &str) -> String {
@@ -83,4 +85,13 @@ pub fn macos_link_env_remove() -> Vec<String> {
     // although this is apparently ignored when using the linker at "/usr/bin/ld".
     env_remove.push("IPHONEOS_DEPLOYMENT_TARGET".to_string());
     env_remove
+}
+
+fn ios_deployment_target() -> (u32, u32) {
+    deployment_target("IPHONEOS_DEPLOYMENT_TARGET").unwrap_or((7, 0))
+}
+
+pub fn ios_sim_llvm_target(arch: &str) -> String {
+    let (major, minor) = ios_deployment_target();
+    format!("{}-apple-ios{}.{}.0-simulator", arch, major, minor)
 }

--- a/compiler/rustc_target/src/spec/apple_sdk_base.rs
+++ b/compiler/rustc_target/src/spec/apple_sdk_base.rs
@@ -11,6 +11,7 @@ pub enum Arch {
     X86_64,
     X86_64_macabi,
     Arm64_macabi,
+    Arm64_sim,
 }
 
 fn target_cpu(arch: Arch) -> String {
@@ -22,13 +23,16 @@ fn target_cpu(arch: Arch) -> String {
         X86_64 => "core2",
         X86_64_macabi => "core2",
         Arm64_macabi => "apple-a12",
+        Arm64_sim => "apple-a12",
     }
     .to_string()
 }
 
 fn link_env_remove(arch: Arch) -> Vec<String> {
     match arch {
-        Armv7 | Armv7s | Arm64 | I386 | X86_64 => vec!["MACOSX_DEPLOYMENT_TARGET".to_string()],
+        Armv7 | Armv7s | Arm64 | I386 | X86_64 | Arm64_sim => {
+            vec!["MACOSX_DEPLOYMENT_TARGET".to_string()]
+        }
         X86_64_macabi | Arm64_macabi => vec!["IPHONEOS_DEPLOYMENT_TARGET".to_string()],
     }
 }

--- a/compiler/rustc_target/src/spec/mod.rs
+++ b/compiler/rustc_target/src/spec/mod.rs
@@ -727,6 +727,7 @@ supported_targets! {
     ("armv7s-apple-ios", armv7s_apple_ios),
     ("x86_64-apple-ios-macabi", x86_64_apple_ios_macabi),
     ("aarch64-apple-ios-macabi", aarch64_apple_ios_macabi),
+    ("aarch64-apple-ios-sim", aarch64_apple_ios_sim),
     ("aarch64-apple-tvos", aarch64_apple_tvos),
     ("x86_64-apple-tvos", x86_64_apple_tvos),
 

--- a/src/doc/rustc/src/platform-support.md
+++ b/src/doc/rustc/src/platform-support.md
@@ -153,6 +153,7 @@ not available.
 target | std | host | notes
 -------|-----|------|-------
 `aarch64-apple-ios-macabi` | ? |  | Apple Catalyst on ARM64
+`aarch64-apple-ios-sim` | ? |  | Apple iOS Simulator on  ARM64
 `aarch64-apple-tvos` | * |  | ARM64 tvOS
 `aarch64-unknown-freebsd` | ✓ | ✓ | ARM64 FreeBSD
 `aarch64-unknown-hermit` | ? |  |

--- a/src/doc/rustc/src/platform-support.md
+++ b/src/doc/rustc/src/platform-support.md
@@ -153,7 +153,7 @@ not available.
 target | std | host | notes
 -------|-----|------|-------
 `aarch64-apple-ios-macabi` | ? |  | Apple Catalyst on ARM64
-`aarch64-apple-ios-sim` | ? |  | Apple iOS Simulator on  ARM64
+`aarch64-apple-ios-sim` | ? |  | Apple iOS Simulator on ARM64
 `aarch64-apple-tvos` | * |  | ARM64 tvOS
 `aarch64-unknown-freebsd` | ✓ | ✓ | ARM64 FreeBSD
 `aarch64-unknown-hermit` | ? |  |

